test/resilience/tests/envoygateway.go

```diff
@@ -0,0 +1,242 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+//go:build resilience
+
+package tests
+
+import (
+	"context"
+	"github.com/envoyproxy/gateway/test/resilience/suite"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/tlog"
+	"testing"
+	"time"
+)
+
+const (
+	namespace    = "envoy-gateway-system"
+	envoygateway = "envoy-gateway"
+	targetString = "successfully acquired lease"
+	apiServerIP  = "10.96.0.1"
+	timeout      = 2 * time.Minute
+	policyName   = "egress-rules"
+	leaseName    = "5b9825d2.gateway.envoyproxy.io"
+	trashHold    = 2
+)
+
+func init() {
+	ResilienceTests = append(ResilienceTests, EGResilience)
+}
+
+var EGResilience = suite.ResilienceTest{
+	ShortName:   "EGResilience",
+	Description: "Envoygateway resilience test",
+	Test: func(t *testing.T, suite *suite.ResilienceTestSuite) {
+		ap := kubernetes.Applier{
+			ManifestFS:     suite.ManifestFS,
+			GatewayClass:   suite.GatewayClassName,
+			ControllerName: "gateway.envoyproxy.io/gatewayclass-controller",
+		}
+		ap.MustApplyWithCleanup(t, suite.Client, suite.TimeoutConfig, "testdata/base.yaml", true)
+
+		//this test will fail until https://github.com/envoyproxy/gateway/pull/4767/files is merged
+		t.Run("Secondary EnvoyGateway instances can serve an up to date xDS", func(t *testing.T) {
+			ctx := context.Background()
+			t.Log("Scaling down the deployment to 0 replicas")
+			err := suite.Kube().ScaleDeploymentAndWait(ctx, envoygateway, namespace, 0, time.Minute, false)
+			require.NoError(t, err, "Failed to scale deployment replicas")
+
+			t.Log("Scaling up the deployment to 3 replicas")
+			err = suite.Kube().ScaleDeploymentAndWait(ctx, envoygateway, namespace, 3, time.Minute, false)
+			require.NoError(t, err, "Failed to scale deployment replicas")
+
+			t.Log("Waiting for leader election")
+			// Ensure leadership was taken
+			name, err := suite.Kube().GetElectedLeader(context.Background(), namespace, leaseName, metav1.Now(), time.Minute*2)
+			require.NoError(t, err, "unable to detect leader election")
+
+			t.Log("Marking the identified pod as leader")
+			suite.Kube().MarkAsLeader(namespace, name)
+
+			t.Log("Simulating API server connection failure for all pods")
```

Comment:
If all pods loose connectivity, don't we expect that they would all fail eventually once becoming leaders?

Reply:
Pods rely on connectivity to the API server to participate in leader election processes. Without this connectivity, they cannot become leaders. Pods should continue to operate using their last known good configuration (xDS) and serve traffic accordingly.